### PR TITLE
Pv2 1023 approvals handle data lock triage approved event

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.Tests.Core/SpecDateExtensions.cs [hunk 1]
```diff
@@ -22,7 +22,11 @@ public static DateTime ToDate(this string dateText)
             }
             return ParseTwoPartDescription(dateText);
         }
-
+        
+        public static DateTime? ToNullableDate(this string dateText)
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.Tests.Core/SpecDateExtensions.cs]: has this been refactored from the ApprenticeshipHelper class? there are two methods that use this code - CreateApprenticeshipModel and CreateApprenticeshipPriceEpisode

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs [hunk 2]
```diff
@@ -15,8 +17,11 @@ public interface IApprenticeshipService
     public class ApprenticeshipService : IApprenticeshipService
     {
         private readonly IApprenticeshipRepository repository;
+      
 
-        public ApprenticeshipService(IApprenticeshipRepository repository)
+        public ApprenticeshipService(IApprenticeshipRepository repository, 
+            IApprenticeshipApprovedUpdatedService apprenticeshipUpdatedService,
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs]: should these be added here as they aren't being used?

File: src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs [hunk 3]
```diff
@@ -93,6 +94,31 @@ public DataLocksProfile()
                 .ForMember(dest => dest.Priority, opt => opt.Ignore())
                 .ForMember(dest => dest.IsLevyPayer, opt => opt.Ignore())
                 ;
+
+            CreateMap<ApprenticeshipUpdatedApprovedEvent, UpdatedApprenticeshipApprovedModel>()
+                .ForMember(dest => dest.ApprenticeshipId, opt => opt.MapFrom(source => source.ApprenticeshipId))
+                .ForMember(dest => dest.StandardCode, opt => opt.ResolveUsing(source => source.TrainingType == ProgrammeType.Standard ? int.Parse(source.TrainingCode) : 0))
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs]: these ResolveUsings are repeats of ones used above. Can they be generalised?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipUpdatedService.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            using (var scope = new TransactionScope(TransactionScopeOption.Required, TransactionScopeAsyncFlowOption.Enabled))
+            {
+
+                HandleUpdated(currentApprenticeship, updatedApprenticeship);
+
+                UpdatePriceEpisodes(updatedApprenticeship.ApprenticeshipPriceEpisodes, currentApprenticeship.ApprenticeshipPriceEpisodes);
+
+                await repository.UpdateApprenticeship(currentApprenticeship);
+
+                var latestApprenticeship = await GetApprenticeship(updatedApprenticeship.ApprenticeshipId).ConfigureAwait(false);
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipUpdatedService.cs]: Why do you need to get the apprenticeship you just saved?

File: src/SFA.DAS.Payments.Tests.Core/SpecDateExtensions.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Tests.Core/SpecDateExtensions.cs]: Didn't want to change existing unless we are working on it.

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipUpdatedService.cs (hunk 4 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipUpdatedService.cs]: We could have added new apprenticeship price episode and we don't know the ID of the new Price Episode so we need to get it back from the DB with the actual Ids before we publish the Apprentice model.

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/ApprenticeshipDataLockTriageServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                .Setup(x => x.Get(It.Is<long>(id => id == apprenticeshipModel.Id)))
+                .ReturnsAsync(apprenticeshipModel);
+
+            var service = mocker.Create<ApprenticeshipDataLockTriageService>();
+            await service.UpdateApprenticeship(updatedApprenticeship);
+
+            mocker.Mock<IApprenticeshipRepository>()
+                .Verify(x => x.UpdateApprenticeship(It.Is<ApprenticeshipModel>(model =>
+                    model.Id == updatedApprenticeship.ApprenticeshipId
+                    && model.AgreedOnDate == updatedApprenticeship.AgreedOnDate
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/ApprenticeshipDataLockTriageServiceTest.cs]: Done

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/ApprenticeshipDataLockTriageServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                        Id =161,
+                        ApprenticeshipId = updatedApprenticeship.ApprenticeshipId,
+                        StartDate = new DateTime(2017,08,10),
+                        Cost = 1000.00m
+                    }
+                }
+            };
+
+            mocker.Mock<IApprenticeshipRepository>()
+                .Setup(x => x.Get(It.Is<long>(id => id == apprenticeshipModel.Id)))
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/ApprenticeshipDataLockTriageServiceTest.cs]: Updated

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs]: Now removed

File: src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs]: Refactored

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/Apprenticeships/ApprenticeshipService.cs]: Removed

**Comment by samuel-dev-dba** [comment]
Closing this PR as this branch has been merged into -- PR Branch
PV2-1024_Approvals_Handle_ApprenticeshipStopped_Event

**Pull request closed without merging**